specreduce/extract.py

```diff
@@ -143,3 +145,108 @@ def _ap_weight_image(trace, width, disp_axis, crossdisp_axis, image_shape):
                           flux=ext1d * getattr(image, 'unit', u.DN))
 
         return spec
+
+
+@dataclass
+class HorneExtract(SpecreduceOperation):
+    """
+    Perform a Horne extraction (a.k.a. optimal) on a region in a
+    two-dimensional spectrum.
+    """
+
+    def __call__(self, image, variance, trace_object,
+                 disp_axis=1, crossdisp_axis=0,
+                 bkgrd_prof=models.Polynomial1D(2)):
+        """
+        Run the Horne calculation on a region of an image and extract a
+        1D spectrum.
+
+        Parameters
+        ----------
+
+        image : `~astropy.nddata.CCDData` or array-like, required
+            The input 2D spectrum from which to extract a source.
+
+        variance : `~astropy.nddata.CCDData` or array-like, required
+            The associated variances for each pixel in the image.
+            Should have the same dimensions as `image`.
+
+        trace_object : `~specreduce.tracing.Trace`, required
+            The associated 1D trace object created for the 2D image.
+
+        disp_axis : int, optional
+            The index of the image's dispersion axis. [default: 1]
+
+        crossdisp_axis : int, optional
+            The index of the image's cross-dispersion axis. [default: 0]
+
+        bkgrd_prof : `~astropy.modeling.Model`
+            A model for the image's background flux.
+            [default: models.Polynomial1D(2)]
+
+        Returns
+        -------
+        spec_1d : `~specutils.Spectrum1D`
+            The final, Horne extracted 1D spectrum.
+        """
+        # co-add signal in each column
+        ncols = image.shape[crossdisp_axis]
+        xd_pixels = np.arange(ncols) # y plot dir / x spec dir
+        coadd = image.sum(axis=disp_axis) / ncols
+
+        # fit source profile, using Gaussian model as a template
+        # NOTE: could add argument for users to provide their own model
+        gauss_prof = models.Gaussian1D(amplitude=coadd.max(),
```

Comment:
Are there any that jump out at you as "must-include?" I mentioned Moffat earlier. Sérsic and Lorentz seem reasonable to me. I'm not sure which are seen as standard in this field.

Reply:
moffat for sure since it's widely applicable for ground-based data. lorentz and voigt would be appropriate as well. sérsic is tricky because it's one-sided. definitely want to stick with symmetric profiles for now...

Reply:
Optimal extraction doesn't work in principle for extended sources.  You can't sensibly weight one part of an extended galaxy more than another.  Optimal extraction is intended for point sources.

Reply:
i was thinking more along the lines of spatially resolved, but compact, galaxies where the spectrum doesn't change significantly as a function of radius. that said, it's a pretty niche case and gaussian + moffat probably cover the vast majority of needs.